components/homme/src/theta-l_kokkos/cxx/Diagnostics.hpp

```diff
@@ -65,37 +66,79 @@ class Diagnostics
 #if ! defined(RESOLVE_ISSUE_WITH_ASSERTS)
-    m_policy(Homme::get_default_team_policy<ExecSpace,EnergyHalfTimesTag>(num_elems)),
+    m_policy_diag_scalars(Homme::get_default_team_policy<ExecSpace,DiagScalarsTag>(num_elems*num_tracers)),
+    m_policy_energy_halftimes(Homme::get_default_team_policy<ExecSpace,EnergyHalfTimesTag>(num_elems)),
 #else
-    m_policy(d_team_policy<EnergyHalfTimesTag>(num_elems)),
+    m_policy_diag_scalars(d_team_policy<DiagScalarsTag>(num_elems*num_tracers)),
+    m_policy_energy_halftimes(d_team_policy<EnergyHalfTimesTag>(num_elems)),
 #endif
-    m_tu(m_policy),
+    m_tu(m_policy_energy_halftimes),
     m_num_elems(num_elems),
+    m_num_tracers(num_tracers),
     m_theta_hydrostatic_mode(theta_hydrostatic_mode)
   {}
 
   void init (const ElementsState& state, const ElementsGeometry& geometry,
-             const HybridVCoord& hvcoord,  F90Ptr& elem_state_q_ptr,
-             F90Ptr& elem_accum_qvar_ptr,  F90Ptr& elem_accum_qmass_ptr,
-             F90Ptr& elem_accum_q1mass_ptr,F90Ptr& elem_accum_iener_ptr,
-             F90Ptr& elem_accum_kener_ptr, F90Ptr& elem_accum_pener_ptr);
+             const HybridVCoord& hvcoord,  const Tracers& tracers,
+             F90Ptr& elem_state_q_ptr, F90Ptr& elem_accum_qvar_ptr,
+             F90Ptr& elem_accum_qmass_ptr, F90Ptr& elem_accum_q1mass_ptr,
+             F90Ptr& elem_accum_iener_ptr, F90Ptr& elem_accum_kener_ptr,
+             F90Ptr& elem_accum_pener_ptr);
 
   int requested_buffer_size () const;
   void init_buffers (const FunctorsBuffersManager& fbm);
 
-  void sync_diags_to_host ();
+  void sync_diagnostics_to_host ();
 
   void run_diagnostics (const bool before_advance, const int ivar);
 
   void prim_diag_scalars (const bool before_advance, const int ivar);
   void prim_energy_halftimes (const bool before_advance, const int ivar);
 
-  HostViewUnmanaged<Real*[QSIZE_D][NUM_PHYSICAL_LEV][NP][NP]> h_Q;
-
+  struct DiagScalarsTag {};
   struct EnergyHalfTimesTag {};
 
+  KOKKOS_INLINE_FUNCTION
+  void operator() (const DiagScalarsTag&, const TeamMember& team) const {
+    KernelVariables kv(team, m_num_tracers);
+
+    Kokkos::parallel_for(Kokkos::TeamThreadRange(kv.team, NP * NP),
+                         [&](const int &point_idx) {
+      const int igp = point_idx / NP;
+      const int jgp = point_idx % NP;
+
+      const auto Q   = viewAsReal(Homme::subview(m_tracers.Q,   kv.ie,         kv.iq, igp, jgp));
+      const auto qdp = viewAsReal(Homme::subview(m_tracers.qdp, kv.ie, t2_qdp, kv.iq, igp, jgp));
+
+      auto& Qvar =   d_Qvar  (kv.ie,m_ivar,kv.iq,igp,jgp);
+      auto& Qmass =  d_Qmass (kv.ie,m_ivar,kv.iq,igp,jgp);
+      auto& Q1mass = d_Q1mass(kv.ie,       kv.iq,igp,jgp);
+
+      // Compute Qvar
+      Qvar = 0.0;
+      Dispatch<>::parallel_reduce(kv.team, Kokkos::ThreadVectorRange(kv.team, NUM_PHYSICAL_LEV),
+                                  [=](const int ilev, Real& accumulator){
```

Comment:
Any reason not to use `[&]` instead of `[=]`, here as well as in the reuctions below?

Reply:
I was just copying what was done in the other operator. But yeah, better to use [&]. I updated and tested.